### PR TITLE
Fix #1232

### LINKED ISSUE #1232: dash-to-dock glitches while in auto-hide setting
When "Auto-hide the Dock" is enabled and there is a maximized window, so that the dock is hidden, upon moving the mouse to the bottom of the screen, sometimes the dock glitches and initiates a hide-show animation without stopping instead from showing up until the cursor moves away from the dock.

https://youtu.be/Lw2HBmLhd6w

This bug was first noticed after upgrading to ubuntu 20.04 from 19.10

Description: Ubuntu 20.04 LTS
Release: 20.04

gnome-shell-extension-ubuntu-dock:
  Installed: 67ubuntu20.04.5
  Candidate: 67ubuntu20.04.5
  Version table:
  67ubuntu20.04.5 500
        500 http://gr.archive.ubuntu.com/ubuntu focal/main amd64 Packages
        500 http://gr.archive.ubuntu.com/ubuntu focal/main i386 Packages
        100 /var/lib/dpkg/status

ProblemType: Bug
ApportVersion: 2.20.11-0ubuntu27
Architecture: amd64
CasperMD5CheckResult: skip
CurrentDesktop: ubuntu:GNOME
DistroRelease: Ubuntu 20.04
InstallationDate: Installed on 2019-10-27 (204 days ago)
InstallationMedia: Ubuntu 19.10 "Eoan Ermine" - Release amd64 (20191017)
NonfreeKernelModules: nvidia_modeset nvidia
Package: gnome-shell-extension-ubuntu-dock 67ubuntu20.04.5
PackageArchitecture: all
ProcVersionSignature: Ubuntu 5.4.0-29.33-generic 5.4.30
Tags: focal
Uname: Linux 5.4.0-29-generic x86_64
UpgradeStatus: Upgraded to focal on 2020-05-09 (9 days ago)
UserGroups: adm cdrom dip lpadmin lxd plugdev sambashare sudo
_MarkForUpload: True

### PR DESCRIPTION
**Pull request by UA27**
Been using this fork for a week now and it fixes #1232 for me. It improves a few odd behaviours around the intellihide/autohide docking setup and mainly deals with the bouncing glitch using `_onPressureSensed`'s pointer-in-dock-area-check method.

On that, I want to state that the main issue came from `_box.hover` being unreliable if any of the maintainers wants to dig into the whole hover ordeal. I mostly worked around it the way I did because I don't know enough about how Gnome handles these things and it made sense to use the same check whatever the dock's state is in to avoid further weird, contradicting and unpredictable behaviours popping up.

_P.S. I was able to reproduce the issue using older commits of dash-to-dock's master on an Ubuntu 20.04 systen, so probably has something to do with newer versions of gnome-shell or I just suck at debugging Gnome, which I do, so feel free to discard anything I said and try it out yourself._

### CODE REVIEW

File: docking.js
Comment (UA27) [inline docking.js:625]: Use `_hide` and `_show` instead to improve readability and keep control over transition behaviours. Only thing is, it now uses the extension's animation delay settings. I don't think it messes with any other setup, but someone with more experience of dash-to-dock's options should probably confirm it.

File: docking.js
Comment (UA27) [inline docking.js:666]: Keeping the `_box.hover` check in an else statement guarantee other docks that do not use `_onPressureSensed` will keep the same behaviour. Because of this, I think setups using intellihide/autohide without the push-to-show option (most likely) will still have the bouncing glitch issue.

Again, I'll leave this part to be reviewed by maintainers with better understanding of the many dash-to-dock's options.

File: docking.js
Comment (UA27) [inline docking.js:676]: Delete the `_delayedHide` value in `_show` instead of `_animateIn` since calls to said `_show` are ignored when the dock's state is either "shown" or "showing" and, when a delayed hide transition is queued, the dock's state is only updated AFTER the animate-in transition is completed.

File: docking.js
Comment (UA27) [inline docking.js:908]: Moved this big boy out so it can be used both in `_onPressureSensed` and `_hoverChanged`.

File: docking.js
Comment (UA27) [inline docking.js:727]: Replace the `_hide` call to `_hoverChanged` the same way `shouldHide` was (is) used in `_onPressureSensed` to catch any `_delayedHide` shenanigans by checking whether the delayed animate-out transition is still necessary instead of forcing the dock to hide right away and trigger a bounce from `_box`'s hover-changed signals.

**Comment by UA27** [comment]
Lastly, this fixes my own issues with #1232, but it probably also concerns #1262, #1251, #1250 and #1207.

**Comment by vltr** [comment]
oh my this issue is **so annoying** and I was almost loosing hope on this matter ... until this proposed fix. I'm using Gnome 3.36 on Arch Linux and the extension seems to be behaving itself "as normal" again! I almost can't believe it ... thank you, @UA27 !

**Comment by GabMus** [comment]
What's blocking this from being merged?

**Comment by k0lourful** [comment]
Thank you for this, helps me as well! I honestly didn't realise this issue is so common until you directed me here.

**Review by 3v1n0 — APPROVED** [review]
Changes looks sane, thanks for looking at this.

**Comment by 3v1n0** [comment]
And thanks for commenting the changes as well!

**Pull request merged**
